packages/monitor-v2/src/llm-bot/OptimisticOracleV2.ts

```diff
@@ -0,0 +1,163 @@
+            requestPriceEvent.args.identifier,
+            requestPriceEvent.args.timestamp,
+            requestPriceEvent.args.ancillaryData
+          )
+          .then((r) => r[4][0]),
+      },
+    });
+    this.requests.set(requestId, newRequest);
+  }
+
+  async applyProposePriceEvent(proposePriceEvent: ProposePriceEvent): Promise<void> {
+    const body = tryHexToUtf8String(proposePriceEvent.args.ancillaryData);
+    const identifier = ethers.utils.parseBytes32String(proposePriceEvent.args.identifier);
+    const timestamp = proposePriceEvent.args.timestamp.toNumber();
+    const requester = proposePriceEvent.args.requester;
+    const requestId = calculateRequestId(body, identifier, timestamp, requester);
+
+    if (!this.requests.has(requestId)) {
+      // A request should always be created before a proposal
+      throw new Error(`Request ${requestId} not found`);
+    }
+    this.requests.set(
+      requestId,
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      this.requests.get(requestId)!.update({
+        proposalData: {
+          proposer: proposePriceEvent.args.proposer,
+          proposedValue: proposePriceEvent.args.proposedPrice,
+          proposeTx: proposePriceEvent.transactionHash,
+          disputableUntil: proposePriceEvent.args.expirationTimestamp.toNumber(),
+        },
+      })
+    );
+  }
+
+  protected async updateOracleRequests(newRange: BlockRange): Promise<void> {
+    const ooV2Contract = await getContractInstanceWithProvider<OptimisticOracleV2Ethers>(
+      "OptimisticOracleV2",
+      this.provider
+    );
+
+    const requestPriceEvents = await this.getEventsWithPagination<RequestPriceEvent>(
+      ooV2Contract.filters.RequestPrice(),
+      newRange[0],
+      newRange[1]
+    );
+
+    const proposePriceEvents = await this.getEventsWithPagination<ProposePriceEvent>(
+      ooV2Contract.filters.ProposePrice(),
+      newRange[0],
+      newRange[1]
+    );
+
+    await Promise.all([
+      ...requestPriceEvents.map(async (requestPriceEvent) => {
+        return this.applyRequestPriceEvent(requestPriceEvent);
+      }),
+    ]);
+
+    await Promise.all([
+      ...proposePriceEvents.map(async (proposePriceEvent) => {
+        return this.applyProposePriceEvent(proposePriceEvent);
+      }),
+    ]);
+  }
+
+  protected createClientInstance(
+    requests: Map<string, OptimisticOracleRequest>,
+    fetchedBlockRange: BlockRange
+  ): OptimisticOracleClientV2 {
+    return new OptimisticOracleClientV2(this.provider, requests, fetchedBlockRange);
+  }
+}
+
+export class OptimisticOracleClientV2FilterDisputeable
+  implements OptimisticOracleClientFilter<OptimisticOracleRequest, OptimisticOracleRequest> {
+  async filter(optimisticOracleRequests: OptimisticOracleRequest[]): Promise<OptimisticOracleRequest[]> {
+    return optimisticOracleRequests.filter((request) => {
+      return request.disputableUntil || 0 > Date.now() / 1000;
```

Comment:
Is it expected by the caller of this filter method that it will also return requests that have not yet been proposed? If it is used by disputer directly then it would revert as you cannot dispute request that has not been proposed